# Exercise 4.4 Data Wrangling & Subsetting

This script contains the following points:
1. Importing libraries
2. Importing data
3. Data wrangling
    3.1 Dropping columns (df_ords)
        3.1.1 Missing values
    3.2 Renaming column (df_ords)
    3.3 Changing data type (df_ords)
    3.4 Transposing data (df_dep)
4. Data Dictionary
5. Subsetting
6. Task 4.4

# 1. Importing libraries

In [44]:
#Import libraries
import pandas as pd
import numpy as np
import os

# 2. Importing data

In [46]:
path = r'C:\Users\Barb\07-2024 Instacart Basket Analysis'

In [47]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original', 'orders.csv'), index_col = False)

In [48]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original', 'products.csv'), index_col = False)

# 3. Data Wrangling

## 3.1 Dropping Column: (df_ords)

In [51]:
#Drop eval_set column from df_ords.
df_ords.drop(columns = ['eval_set'])

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [52]:
#Overwrite df_ords to reflect missing eval_set column
df_ords = df_ords.drop(columns = ['eval_set'])

### 3.1.1 Find Missing Values

In [54]:
#Find missing values in the 'days_since_prior_order' column
df_ords['days_since_prior_order'].value_counts(dropna = False)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

## 3.2 Renaming column: (df_ords)

In [56]:
#Rename 'order_dow' column
df_ords.rename(columns = {'order_dow': 'orders_day_of_week'}, inplace = True)

In [57]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

## 3.3 Changing data type: (df_ords)

In [59]:
#change data type of 'order_id' and 'user_id' to a string (or object) - a data type the command decribe() will ignore
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [60]:
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [62]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 6 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                object 
 1   user_id                 object 
 2   order_number            int64  
 3   orders_day_of_week      int64  
 4   order_hour_of_day       int64  
 5   days_since_prior_order  float64
dtypes: float64(1), int64(3), object(2)
memory usage: 156.6+ MB


## 3.4 Transposing data: (df_dep)

In [64]:
#Importing data set departments.csv
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original', 'departments.csv'), index_col = False)

In [65]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [66]:
# Transposing df_def
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [67]:
# Overwrite transposed version
df_dep_t = df_dep.T

In [68]:
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [69]:
# Add index to  df_dep_t dataframe
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [70]:
#Take the first row of df_dip_t for the header
new_header = df_dep_t.iloc[0]

In [71]:
new_header

0    department
Name: department_id, dtype: object

In [72]:
# Take the data under the header row for a new df
df_dep_t_new = df_dep_t[1:]

In [73]:
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [74]:
df_dep_t_new.columns = new_header #set the header row as the df header

In [75]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

# 4. Data Dictionary

In [77]:
# Turn your df_dep_t_new dataframe into a data dictionary
data_dict = df_dep_t_new.to_dict('index')

In [78]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [79]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [80]:
print(data_dict.get('19'))

{'department': 'snacks'}


# 5. Subsetting

In [82]:
# Overall Goal- 3 Steps -Create a subset for df_prods dataframe that only contains data from the snacks department
# Final code: df_snacks = df_prods[df_prods['department_id']==19]

In [83]:
# Step 1 - Run the indexing code. This simply searches for the data in question: everything with a "department_id" of 19
df_prods['department_id']==19

0         True
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool

In [84]:
# Step 2- Enclosing step #1 inside df_prods[] tells Python to map above true/false procedure from step #1 onto columns w/i the df_prods df
#          resulting inn a list of only those values w/i df_prods that have a "department_id" of 19.
df_prods[df_prods['department_id']==19]

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

In [85]:
# Step 3- Save the results of step 2 into new dataframe df_snacks
df_snacks = df_prods[df_prods['department_id']==19]

In [86]:
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

# 6. Task 4.4: Problems #2-13

#### Task #2. user_id in df_ords doesn't need to be included in your analysis as a numerical variable, so change it to a string (object)

In [89]:
# Change data type of "user_id" column from a number to a string.
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [90]:
# Check that the data type of 'user_id' changed.
df_ords['user_id'].dtype

dtype('O')

#### Task #3a. Look for a variable in the df_ords df w/ an unintuitive name and change its name without overwriting the df

In [92]:
df_ords.head()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329       1             1                   2                  8   
1  2398795       1             2                   3                  7   
2   473747       1             3                   3                 12   
3  2254736       1             4                   4                  7   
4   431534       1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

#### Task #3b. Rename "order_hour_of_day' order w/o rewriting it

In [94]:
df_ords.rename(columns = {'order_hour_of_day': 'hour_time_of_order'}, inplace = True)

In [95]:
df_ords.head()

order_id user_id  order_number  orders_day_of_week  hour_time_of_order  \
0  2539329       1             1                   2                   8   
1  2398795       1             2                   3                   7   
2   473747       1             3                   3                  12   
3  2254736       1             4                   4                   7   
4   431534       1             5                   4                  15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

#### Task #4 - What is the busiest hour for placing orders? Find the frequency of the corresponding variable

In [97]:
df_ords['hour_time_of_order'].value_counts(dropna = True)

hour_time_of_order
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

### #4 Answer: 10 AM is the busiest hour for placing orders.

Observation: Even though I did not rewrite the renaming of the column 'order_hour_of_day', I had to use the new name 'hour_time_of_order'.
Otherwise, I got an error message if I used 'order_hour_of_day.

#### Task #5 - Find the meaning behind a value of 4 in the "department_id" column w/i the df_prods df using a data dictionary

In [101]:
print(data_dict.get('4'))

{'department': 'produce'}


### #5 Answer - Department ID of 4 means "produce"

#### Task #6 Create a subset containing only info about breakfast item sales.

In [104]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [105]:
df_breakfast = df_prods[df_prods['department_id']==14]

In [106]:
df_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

#### Task #7 Find all observations from the entire dataframe that include items from the following departments:
####       alcohol, deli, beverages, and meat/seafood

In [108]:
df_dinner_party = df_prods.loc[df_prods['department_id'].isin([5,7,12,20])]

In [109]:
df_dinner_party.head()

product_id                                    product_name  aisle_id  \
2            3            Robust Golden Unsweetened Oolong Tea        94   
6            7                  Pure Coconut Water With Orange        98   
9           10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                               Peach Mango Juice        31   
16          17                               Rendered Duck Fat        35   

    department_id  prices  
2               7     4.5  
6               7     4.4  
9               7     8.4  
10              7     2.8  
16             12    17.1

#### Task #8 How many rows from the last datafrome?

In [111]:
df_dinner_party.shape

(7650, 5)

### #8 Answer: 7650 rows in df_dinner_party

#### Task #9 Extract all the info you can about customer with "user_id" of '1'.

In [114]:
#Create a subset df_user1 containing only info about customer with "user_id" of '1'.

In [115]:
df_user1 = df_ords[df_ords['user_id']=='1']

In [116]:
df_user1.head()

order_id user_id  order_number  orders_day_of_week  hour_time_of_order  \
0  2539329       1             1                   2                   8   
1  2398795       1             2                   3                   7   
2   473747       1             3                   3                  12   
3  2254736       1             4                   4                   7   
4   431534       1             5                   4                  15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [117]:
#Find rows from df_user1
df_user1.shape

(11, 6)

In [118]:
df_user1.head(11)

order_id user_id  order_number  orders_day_of_week  hour_time_of_order  \
0   2539329       1             1                   2                   8   
1   2398795       1             2                   3                   7   
2    473747       1             3                   3                  12   
3   2254736       1             4                   4                   7   
4    431534       1             5                   4                  15   
5   3367565       1             6                   2                   7   
6    550135       1             7                   1                   9   
7   3108588       1             8                   1                  14   
8   2295261       1             9                   1                  16   
9   2550362       1            10                   4                   8   
10  1187899       1            11                   4                   8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

#### Task #10 Provide some details about behavior of user_id = 1. What basic stats can you provide based on the info you have?

In [120]:
# Find out what numbers are used for orders_day_of_week

In [121]:
df_ords['orders_day_of_week']. value_counts(dropna = False)

orders_day_of_week
0    600905
1    587478
2    467260
5    453368
6    448761
3    436972
4    426339
Name: count, dtype: int64

In [122]:
# Generate descriptive statistics
df_user1.describe()

order_number  orders_day_of_week  hour_time_of_order  \
count     11.000000           11.000000           11.000000   
mean       6.000000            2.636364           10.090909   
std        3.316625            1.286291            3.477198   
min        1.000000            1.000000            7.000000   
25%        3.500000            1.500000            7.500000   
50%        6.000000            3.000000            8.000000   
75%        8.500000            4.000000           13.000000   
max       11.000000            4.000000           16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

### #10 Answer: User 1 made 11 orders total.
###             1/2 of the orders were before 8 am.
###             He made orders on Sun - Wed. For 0 = Saturday
###             Max time between orders was 30 days.
###             Customer ordered twice on the same day (Sunday).

## 6.1 Exporting Dataframe

#### Task #12 Export your df_ords dataframe as "orders_wrangled.csv" in your "Prepared Data" folder

In [126]:
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared', 'orders_wrangled.csv'))

#### Task #13. Export the df_dep_t_new dataframe as "departments_wrangled.csv" in your "Prepared Data" folder

In [128]:
df_dep_t_new.to_csv(os.path.join(path, '02 Data', 'Prepared', 'departments_wrangled.csv'))